### PR TITLE
Network Strategies

### PR DESCRIPTION
**Pull request by AndrewBastin**
This PR intends to introduce Network Strategies into Postwoman.

Network Strategies are basically functions which take in the Postwoman Store and the request config and returns a Promise for the response. It generalizes the network request making process and makes further introduction of features easier.

Currently, this PR introduces 2 Network Strategies

- **AxiosStrategy** : Uses axios to send the request
- **ProxyStrategy** : Uses the postwoman proxy feature to send the request through a proxy.

Currently, Network Strategies allow its consumers to not worry about whether proxy is enabled, the `sendNetworkRequest` (found in functions/network.js) function will detect the appropriate strategy from the available and executes that. But in the future, when we plan to add new strategies (for e.g for native), we do not need to worry about rewriting the consumers to adopt those.

### CODE REVIEW

**Comment by liyasthomas** [comment]
Axios progress bar seems to be broken.

To reproduce:
1. Visit deploy preview: https://deploy-preview-487--postwoman.netlify.com/
2. Send a request
3. Progress won't finish even after axios fetch response.

Except this, I couldn't find any bugs.

**Comment by AndrewBastin** [comment]
Well, the AxiosStrategy doesn't use the `$axios` plugin thingy. Does the progress bar depend on that or something ? If so, we will need to fix that.

**Comment by liyasthomas** [comment]
Well axios automatically binds progress to Nuxt's progress bar. We can turn that off, and invoke progress manually. Lemme find a fix for that.

File: functions/network.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import AxiosStrategy from "./strategies/AxiosStrategy";
+import ProxyStrategy from "./strategies/ProxyStrategy";
+
+
+const sendNetworkRequest = (req, store) => {
+  if (store.state.postwoman.settings.PROXY_ENABLED) {
+    return ProxyStrategy(req, store);
+  } else {
+    return AxiosStrategy(req, store);
+  }
```
Comment (jamesgeorge007) [inline functions/network.js]: Avoid the `else` block.

```js
  if (store.state.postwoman.settings.PROXY_ENABLED) {
    return ProxyStrategy(req, store);
  }
  return AxiosStrategy(req, store);
```

File: functions/strategies/ProxyStrategy.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import axios from "axios";
+
+const proxyStrategy = async (req, store) => {
+  const proxyRes = await axios.post(
```
Comment (jamesgeorge007) [inline functions/strategies/ProxyStrategy.js]:
```suggestion
  const { data } = await axios.post(
```

File: functions/strategies/ProxyStrategy.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+    store.state.postwoman.settings.PROXY_URL ||
+    "https://postwoman.apollotv.xyz/",
+    req
+  );
+
+  return proxyRes.data;
```
Comment (jamesgeorge007) [inline functions/strategies/ProxyStrategy.js]:
```suggestion
  return data;
```

File: functions/strategies/ProxyStrategy.js (hunk 2 above)
Comment (jamesgeorge007) [reply functions/strategies/ProxyStrategy.js]: I'm a big fan of the object destructuring approach. It's quite subjective :sweat_smile:

File: functions/strategies/ProxyStrategy.js (hunk 3 above)
Comment (AndrewBastin) [reply functions/strategies/ProxyStrategy.js]: Updated in the latest commit

File: functions/strategies/ProxyStrategy.js (hunk 2 above)
Comment (AndrewBastin) [reply functions/strategies/ProxyStrategy.js]: Updated in the latest commit

File: functions/network.js (hunk 1 above)
Comment (AndrewBastin) [reply functions/network.js]: Updated in the latest commit

**Comment by AndrewBastin** [comment]
@liyasthomas what about the progress bar issue ?

**Comment by liyasthomas** [comment]
Fix coming in few mins

**Pull request merged**